File: kernel/asm/int.rs
#![feature(asm)]

static mut CLI_COUNT: u32 = 1;

#[no_mangle]
pub fn cli() {
    unsafe {
        asm!("cli");
        CLI_COUNT += 1;
    }
}

#[no_mangle]
pub fn sti() {
    unsafe {
        CLI_COUNT -= 1;
        if CLI_COUNT == 0 {
            asm!("sti");
        }
    }
}

File: kernel/asm/int.c
int cli_count = 1;

void cli() {
    __asm__ volatile ("cli");
    cli_count++;
}

void sti() {
    cli_count--;
    if (cli_count == 0) {
        __asm__ volatile ("sti");
    }
}

void hlt() {
    __asm__ volatile ("hlt");
}
